crates/core/tedge_mapper/src/mapper.rs

```diff
@@ -85,14 +89,121 @@ impl Mapper {
+
+        loop {
+            tokio::select! {
+                // When the reconciliation window times out, break out of the select loop to stop caching alarm messages
+                _ = &mut reconciliation_window => {
+                    break;
+                }
+
+                // During the reconciliation window, cache all the alarm messages received from tedge/alarms and tedge-internal alarms topics.
+                // This window gives the mapper ample time to receive all the retained messages from both the topics.
+                Some(message) = self.input.next() => {
+                    if message.topic.name.starts_with("tedge/alarms") {
+                        let alarm_id = message
+                            .topic
+                            .name
+                            .strip_prefix("tedge/alarms/")
+                            .expect("Expected tedge/alarms prefix")
+                            .to_string();
+                        pending_alarms_map
+                            .insert(alarm_id.clone(), message);
+                    } else if message.topic.name.starts_with("tedge-internal/alarms") {
+                        let alarm_id = message
+                            .topic
+                            .name
+                            .strip_prefix("tedge-internal/alarms/")
+                            .expect("Expected tedge-internal/alarms prefix")
+                            .to_string();
+                        old_alarms_map
+                            .insert(alarm_id, message);
+                    }
+                }
+            }
+        }
+
+        // Compare the differences between alarms in tedge/alarms topic to the ones in tedge-internal/alarms topic
+        for (key, old_message) in old_alarms_map.into_iter() {
+            match pending_alarms_map.get(&key) {
+                // If an alarm that is present in tedge-internal/alarms topic is not present in tedge/alarms topic,
+                // it is assumed to have been cleared while the mapper process was down
+                None => {
+                    let topic = Topic::new_unchecked(format!("tedge/alarms/{}", key).as_str());
+                    let message = Message::new(&topic, vec![]).with_retain();
+                    // Recreate the clear alarm message and add it to the pending alarms list to be processed later
+                    pending_alarms_map.insert(key, message);
+                }
+
+                // If the payload of a message received from tedge/alarms is same as one received from tedge-internal/alarms,
+                // it is assumed to be one that was already processed earlier and hence removed from the pending alarms list.
+                Some(message) => {
+                    if message.payload_bytes() == old_message.payload_bytes() {
+                        pending_alarms_map.remove(&key);
+                    }
+                }
+            }
+        }
```

Comment:
An alternative implementation:
```suggestion
        old_alarms_map.iter().for_each(|(alarm_id, message)| {
            match pending_alarms_map.entry(alarm_id.to_string()) {
                Entry::Occupied(e) => {
                    if message.payload_bytes() == e.get().payload_bytes() {
                        e.remove();
                    }
                }

                Entry::Vacant(e) => {
                    let topic = Topic::new_unchecked(format!("tedge/alarms/{}", e.key()).as_str());
                    let message = Message::new(&topic, vec![]).with_retain();
                    e.insert(message);
                }
            }
        });
```

I hope I got the logic right.